lib/ecto_ordered.ex

```diff
@@ -65,190 +58,270 @@ defmodule EctoOrdered do
-    query = from m in module,
-    where: field(m, ^field) > ^split_by and field(m, ^field) <= ^until
-            and field(m, ^scope) == ^old_scope
-    execute_decrement(struct, query)
+  defp rank_for_row(old_rank, index, count, old_attempted_rank) do
+    # If our old rank is less than the old attempted rank, then our effective index is fine
+    new_index = if old_rank < old_attempted_rank do
+      index
+    # otherwise, we need to increment our index by 1
+    else
+      index + 1
+    end
+    round((@max - @min) / count) * new_index + @min
   end
 
-  defp validate_position!(cs, %Order{field: field, new_position: position, max: max}) when position > max + 1 do
-    raise EctoOrdered.InvalidMove, type: :too_large
-    %Ecto.Changeset{ cs | valid?: false } |> add_error(field, :too_large)
-  end
-  defp validate_position!(cs, %Order{field: field, new_position: position}) when position < 1 do
-    raise EctoOrdered.InvalidMove, type: :too_small
-    %Ecto.Changeset{ cs | valid?: false } |> add_error(field, :too_small)
+  defp current_order(%Order{rank_field: rank_field, repo: repo} = order, cs) do
+    order
+    |> queryable
+    |> ranked(rank_field)
+    |> scope_query(order, cs)
+    |> repo.all
   end
-  defp validate_position!(cs, _), do: cs
 
-  defp update_old_scope(%Order{scope: scope} = struct, cs) do
-    %{struct|old_scope: Map.get(cs.model, scope)}
+  defp shift_others_up(%Order{rank_field: rank_field,
+                              repo: repo} = order, %{model: existing} = cs) do
+    current_rank = get_field(cs, rank_field)
+    order
+    |> queryable
+    |> where([r], field(r, ^rank_field) >= ^current_rank)
+    |> exclude_existing(existing)
+    |> repo.update_all([inc: [{rank_field, 1}]])
+    cs
   end
 
-  defp update_new_scope(%Order{scope: scope} = struct, cs) do
-    %{struct|new_scope: get_field(cs, scope)}
+  defp shift_others_down(%Order{rank_field: rank_field,
+                                repo: repo} = order, %{model: existing} = cs) do
+    current_rank = get_field(cs, rank_field)
+    order
+    |> queryable
+    |> where([r], field(r, ^rank_field) <= ^current_rank)
+    |> exclude_existing(existing)
+    |> repo.update_all([inc: [{rank_field, -1}]])
+    cs
   end
 
-  defp update_new_position(%Order{field: field} = struct, cs) do
-    %{struct|new_position: get_field(cs, field)}
+  defp current_at_rank(%Order{repo: repo, rank_field: rank_field} = order, cs) do
+    rank = get_field(cs, rank_field)
+    order
+    |> queryable
+    |> where([r], field(r, ^rank_field) == ^rank)
+    |> limit(1)
+    |> scope_query(order, cs)
+    |> repo.one
   end
 
-  defp update_old_position(%Order{field: field} = struct, cs) do
-    %{struct|old_position: Map.get(cs.model, field)}
+  defp neighbours_at_position(%Order{
+                                     rank_field: rank_field,
+                                     repo: repo
+                                    } = order, position, cs) when position <= 0 do
```

Comment:
Yeah. Credo doesn't bother me about it, but it would bother me about super-long lines.

In some of these places, I don't need to match in the function definition, but I was just following on the style from the existing library.